### PR TITLE
Qualcomm AI Engine Direct - Pass migration - part 1

### PR DESCRIPTION
**Pull request by zhaoxul-qti**
## Summary

This is **part 1** of a series whose goal is to deprecate `to_edge_transform_and_lower_to_qnn` and migrate Qualcomm-specific passes from the export pipeline into `to_edge_transform_and_lower`.

The core change is in `exir/program/_program.py`: after the `EdgeProgramManager` finishes running `transform_passes`, we now invoke `lift_constant_tensor_pass` on each method's `ExportedProgram`. Several QNN passes (`CanonicalizeConv`, `ConvertLinearToConv2d`, etc.) rewrite constant values and insert new `get_attr` nodes to store them. Without lifting, these mutated constants are not reflected back into the `ExportedProgram`; `lift_constant_tensor_pass` promotes the `get_attr` nodes into placeholders so the modified constants are correctly propagated.

## Changes
- `exir/program/_program.py`: run `lift_constant_tensor_pass` over every method after `edge_manager.transform(transform_passes)` completes.
- Move `CanonicalizeConv` and `ConvertLinearToConv2d` out of the export pipeline (`get_export_passes`) and into the to-edge transform pipeline.
- Add new pass `AnnotateGetAttr` to repopulate `QCOM_QUANT_ATTRS` for `get_attr` nodes. Passes such as `I64toI32` and `LayoutTransform` reconstruct the `GraphModule` (`graph_module = super().call(graph_module).graph_module`), which drops the quant attributes previously stored on `get_attr` node metadata.

## Test plan
- `CanonicalizeConv`
Test cases:
`TEST_MATRIX = {TestQNNFloatingPointOperator.test_qnn_backend_conv1d TestQNNFloatingPointOperator.test_qnn_conv1d_batch_norm TestQNNFloatingPointOperator.test_qnn_backend_conv2d TestQNNFloatingPointOperator.test_qnn_backend_conv3d_sequential TestQNNFloatingPointOperator.test_qnn_backend_conv_transpose1d TestQNNFloatingPointOperator.test_qnn_backend_conv_transpose2d TestQNNFloatingPointOperator.test_qnn_backend_conv_transpose3d TestQNNFloatingPointModel.test_qnn_backend_conv1d_relu_log_softmax TestQNNQuantizedOperator.test_qnn_backend_conv1d TestQNNQuantizedOperator.test_qnn_conv1d_batch_norm TestQNNQuantizedOperator.test_qnn_backend_conv2d TestQNNQuantizedOperator.test_qnn_backend_conv3d_sequential TestQNNQuantizedOperator.test_qnn_backend_conv_transpose1d TestQNNQuantizedOperator.test_qnn_backend_conv_transpose2d TestQNNQuantizedOperator.test_qnn_backend_conv_transpose3d TestQNNQuantizedModel.test_qnn_backend_conv1d_relu_log_softmax}`
Command:
`python backends/qualcomm/tests/test_qnn_delegate.py ${TEST_MATRIX} --build_folder build-android/ --host ${HOST_NAME} --device ${DEVICE_ID} --soc_model ${SOC_ID} --seed 1126 --backend htp`
`python -m pytest "backends/qualcomm/tests/rework/passes/test.py::test_canonicalize_conv[htp]" -v`
`python -m pytest "backends/qualcomm/tests/rework/passes/test.py::test_canonicalize_conv[lpai]" -v`

- `ConvertLinearToConv2d`
Test cases:
`TEST_MATRIX = {TestQNNFloatingPointOperator.test_qnn_backend_linear TestQNNFloatingPointOperator.test_qnn_backend_linear_to_conv2d TestQNNFloatingPointOperator.test_qnn_backend_linear_shared_weights TestQNNFloatingPointOperator.test_qnn_backend_linear_to_conv2d_shared_weights  TestQNNQuantizedOperator.test_qnn_backend_linear TestQNNQuantizedOperator.test_qnn_backend_linear_to_conv2d TestQNNQuantizedOperator.test_qnn_backend_linear_shared_weights TestQNNQuantizedOperator.test_qnn_backend_linear_to_conv2d_shared_weights TestQNNQuantizedOperator.test_qnn_backend_linear_block TestQNNQuantizedOperator.test_qnn_backend_linear_to_conv2d_block TestQNNQuantizedOperator.test_qnn_backend_linear_qat}`
Command:
`python backends/qualcomm/tests/test_qnn_delegate.py ${TEST_MATRIX} --build_folder build-android/ --host ${HOST_NAME} --device ${DEVICE_ID} --soc_model ${SOC_ID} --seed 1126 --backend htp`
`python -m pytest "backends/qualcomm/tests/rework/passes/test.py::test_convert_linear_to_conv2d[htp]" -v`
`python -m pytest "backends/qualcomm/tests/rework/passes/test.py::test_convert_linear_to_conv2d[lpai]" -v`
`python -m pytest "backends/qualcomm/tests/rework/htp/op/v68/test.py::test_linear_shared_weights" -v`

- `AnnotateGetAttr`
Command:
`python -m pytest "backends/qualcomm/tests/rework/passes/test.py::test_annotate_get_attr" -v`

### CODE REVIEW

**Comment by zhaoxul-qti** [comment]
Hi @psiddh,
This PR is part one of the series to deprecate the `to_edge_transform_and_lower_to_qnn`.
Would you mind take a look on this PR?
Thanks for the review on this!

File: exir/program/_program.py [hunk 1]
```diff
@@ -1344,6 +1344,8 @@ def to_edge_transform_and_lower(  # noqa: C901
 
     if transform_passes is not None:
         edge_manager = edge_manager.transform(transform_passes)
+        for method in edge_manager.methods:
+            lift_constant_tensor_pass(edge_manager.exported_program(method))
```
Comment (psiddh) [inline exir/program/_program.py:1346]: This is the shared export code  that every backend goes through, not just specific to QNN. wouldn't it silently change the other backend's input signatures? Instead of here can we restrict it into QNN specific flow ?

File: backends/qualcomm/_passes/annotate_get_attr.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    def __init__(self, edge_program: torch.export.ExportedProgram):
+        self.edge_program = edge_program
+
+    def _annotate_get_attr(
+        self, graph_module: torch.fx.GraphModule
+    ) -> torch.fx.GraphModule:
+        for node in graph_module.graph.nodes:
+            if node.op == "get_attr" and list(node.users)[0].target in dq_ops:
+                dq_op = list(node.users)[0]
```
Comment (psiddh) [inline backends/qualcomm/_passes/annotate_get_attr.py]: Is it possible that get_attr has no users  ? If so it'll IndexError and abort the export ? Also since it only checks the first user (is this guaranteed ), a constant whose dq isn't first would silently miss other attrs? Could we guard on node.users and scan for any dq user instead of assuming [0] ?

File: backends/qualcomm/_passes/annotate_get_attr.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+import torch
+from executorch.backends.qualcomm.builders.node_visitor import dq_ops
+from executorch.backends.qualcomm.utils.constants import QCOM_QUANT_ATTRS
+from executorch.exir.pass_base import ExportPass, PassResult
+
+from .utils import get_quant_attrs
+
+
+class AnnotateGetAttr(ExportPass):
```
Comment (psiddh) [inline backends/qualcomm/_passes/annotate_get_attr.py:15]: should it call super().__init__()

Comment (psiddh) [inline backends/qualcomm/_passes/annotate_get_attr.py:15]: Also this pass exists because other passes drop quant metadata on get_attr nodes ? If so, would it be better to have other passes preserve the it instead ? What do you think ?

**Comment by psiddh** [comment]
@claude  review

File: exir/program/_program.py (hunk 1 above)
Comment (zhaoxul-qti) [reply exir/program/_program.py:1346]: This is the most critical part of the entire migration series.

One of the main challenges is that `transform_passes` currently operates on a `GraphModule` rather than an `ExportedProgram`. As a result, any pass that needs to modify constants at this stage (for example, converting Conv1d to Conv2d, which requires updating the weight tensor and propagating the change back to the `ExportedProgram`'s input signature) cannot be handled correctly without lifting the constant tensors. This is the reason `lift_constant_tensor_pass` is required here.

In the original workflow, this logic was confined to the QNN-specific path within `to_edge_transform_and_lower_to_qnn`, which is exactly the flow that this PR is attempting to deprecate and migrate away from.

Given that this is the shared export code that every backend goes through, I was wondering if these two lines of change be considered acceptable once the full CI suite passes across all backends.

File: backends/qualcomm/_passes/annotate_get_attr.py (hunk 2 above)
Comment (zhaoxul-qti) [reply backends/qualcomm/_passes/annotate_get_attr.py]: Thanks for pointing this out.

I don't expect the `get_attr` node to have no users in this case, but I agree that adding a defensive check would make it more robust. I'll also add the check and validate for any dq user in `node.users`.

File: backends/qualcomm/_passes/annotate_get_attr.py (hunk 3 above)
Comment (zhaoxul-qti) [reply backends/qualcomm/_passes/annotate_get_attr.py:15]: I don't think this is required, as the current implementation works correctly without it.

Comment (zhaoxul-qti) [reply backends/qualcomm/_passes/annotate_get_attr.py:15]: Yes, this pass exists because other passes can inadvertently drop quant metadata on `get_attr` nodes.

Although we could patch each affected pass individually, that approach does not scale well. Any future pass that recreates or transforms the graph through

    graph_module = super().call(graph_module).graph_module

would need to remember to preserve the quant attributes for `get_attr` nodes as well. Missing that step could easily introduce regressions.

For that reason, I added a generic `AnnotateGetAttr` pass to address this problem once for all. I believe this provides a more uniform and maintainable solution, while also reducing the risk of future passes accidentally breaking the `get_attr`'s quantization metadata propagation.

File: exir/program/_program.py (hunk 1 above)
Comment (zhaoxul-qti) [reply exir/program/_program.py:1346]: Hi @psiddh,
I'd appreciate your perspective on this part when you have a chance.

**Comment by psiddh** [comment]
@zhaoxul-qti can you take a look at the comments above ?

**Comment by shewu-quic** [comment]
> @zhaoxul-qti can you take a look at the comments above ?

Sorry, @zhaoxul-qti is currently OoO and should be back this week. He'll take a look at the comments and respond as soon as possible.

Thanks for the review!

**Comment by zhaoxul-qti** [comment]
Hi @psiddh,
I have addressed all the comments above and pushed a new commit on it.
I think this PR is ready to merge once there is no CI failure.
Thanks for reviewing it.

**Review by psiddh — APPROVED** [review]
